File: 0077-combinations/0077-combinations.java
class Solution {
    private void dfs(List<List<Integer>> results, LinkedList<Integer> elements, int n, int start, int k) {
        if (k == 0) {
            results.add(elements.stream().collect(Collectors.toList()));
            return;
        }

        for (int i = start; i <= n; i++) {
            elements.add(i);
            dfs(results, elements, n, i + 1, k - 1);
            elements.removeLast();
        }
    }

    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> results = new ArrayList<>();
        dfs(results, new LinkedList<>(), n, 1, k);

        return results;
    }
}

File: 0077-combinations/0077-combinations.kt
class Solution {
    fun combine(n: Int, k: Int): List<List<Int>> {
        val results: MutableList<List<Int>> = mutableListOf()

        fun dfs(elements: LinkedList<Int>, start: Int, k: Int) {
            if (k == 0) {
                results.add(elements.stream().collect(Collectors.toList()))
                return
            }

            for (i in start..n) {
                elements.add(i)
                dfs(elements, i + 1, k - 1)
                elements.removeLast()
            }
        }

        dfs(LinkedList(), 1, k)

        return results
    }
}